cmd/askd/handlers/form.go

```diff
@@ -102,3 +107,216 @@ func (formHandle) Delete(c *web.Context) error {
+type FormDigest struct {
+	Questions map[string]FormQuestionDigest `json:"questions" bson:"questions"`
+}
+
+// Digest returns a form digest.
+// 200 Success, 400 Bad Request, 404 Not Found, 500 Internal
+func (formHandle) Digest(c *web.Context) error {
+	id := c.Params["form_id"]
+
+	// Load the form requested.
+	f, err := form.Retrieve(c.SessionID, c.Ctx["DB"].(*db.DB), id)
+	if err != nil {
+		return err
+	}
+
+	// Create a container for the question digests.
+	questions := make(map[string]FormQuestionDigest)
+
+	// Order is a counter to set the order questions in the form.
+	order := 1
+
+	// Loop through to form's steps/widgets to find the questions.
+	for _, step := range f.Steps {
+		for _, widget := range step.Widgets {
+
+			// Unpack the question properties.
+			props := widget.Props.(bson.M)
+
+			// We are looking to only include submissions with includeInGroups or groupSubmissions.
+			gs, gsok := props["groupSubmissions"]
+			iig, iigok := props["includeInGroups"]
+
+			// Skip other questions, and do it verbosely to protect against messy data.
+			if (gs == nil || !gsok || gs == false) && (iig == nil || !iigok || iig == false) {
+				continue
+			}
+
+			// Include options for MultipleChoice questions.
+			options := []FormQuestionOptionDigest{}
+			if widget.Component == "MultipleChoice" {
+
+				// Step outside the safety of the type system...
+				opts := props["options"].([]interface{})
+
+				for _, opt := range opts {
+					option := opt.(bson.M)
+
+					// Hash the answer text for a unique key, as no actual key exists.
+					hasher := md5.New()
+					hasher.Write([]byte(option["title"].(string)))
+					optKeyStr := hex.EncodeToString(hasher.Sum(nil))
+
+					// Add this option to the array.
+					options = append(options, FormQuestionOptionDigest{
+						ID:    optKeyStr,
+						Value: option["title"].(string),
+					})
+				}
+			}
+
+			// Add the question to the digest.
+			questions[widget.ID] = FormQuestionDigest{
+				ID:      widget.ID,
+				Type:    widget.Component,
+				Title:   widget.Title,
+				GroupBy: gsok,
+				Options: options,
+				Order:   order,
+			}
+
+			// Increment the order counter.
+			order = order + 1
+
+		}
+	}
+
+	digest := FormDigest{
+		Questions: questions,
+	}
+
+	c.Respond(digest, http.StatusOK)
+
+	return nil
+}
+
+//==============================================================================
+
+// Submission Groups: The submission answers marked with includeInGroups from
+// the given groups, or all.
+
+//
+// 200 Success, 400 Bad Request, 404 Not Found, 500 Internal
+func (formHandle) SubmissionGroup(c *web.Context) error {
+	id := c.Params["form_id"]
+
+	// unpack the search headers and create a SearchOpts for Group Submissions
+
+	limit, err := strconv.Atoi(c.Request.URL.Query().Get("limit"))
+	if err != nil {
+		limit = 0
+	}
+
+	skip, err := strconv.Atoi(c.Request.URL.Query().Get("skip"))
+	if err != nil {
+		skip = 0
+	}
+
+	opts := submission.SearchOpts{
```

Comment:
Need some comments throughout this function.  I'm having trouble following intent.